packages/core/src/RenderingEngine/helpers/getOrCreateCanvas.ts

```diff
@@ -54,5 +63,26 @@ export default function getOrCreateCanvas(
   const internalDiv =
     element.querySelector(viewportElement) || createViewportElement(element);
 
-  return internalDiv.querySelector(canvasSelector) || createCanvas(internalDiv);
+  const canvas = (internalDiv.querySelector(canvasSelector) ||
+    createCanvas(internalDiv)) as HTMLCanvasElement;
+  // Fit the canvas into the div
+  const rect = internalDiv.getBoundingClientRect();
+  const devicePixelRatio = window.devicePixelRatio || 1;
+
+  // The width/height is the number of physical pixels which will completely
+  // cover the div so that no pixels, fractional or full are left uncovered.
+  // Thus, it is the ceiling of the CSS size times the physical pixels.
+  // In theory, the physical pixels can be offset from CSS pixels, but in practice
+  // this hasn't been observed.
+  const width = Math.ceil(rect.width * devicePixelRatio);
+  const height = Math.ceil(rect.height * devicePixelRatio);
+  canvas.width = width;
+  canvas.height = height;
```

Comment:
I expected some left and top manipulation as per our discussion, howcome the width and height ceiling handles making the green canvas below

![CleanShot 2023-09-11 at 22 19 24](https://github.com/cornerstonejs/cornerstone3D/assets/7490180/5e479163-0c97-49a7-945c-5362018d2014)

Reply:
That is because of your comment about leaving a tiny bit of unfilled background.  What I did was set it up so that there is just more than enough canvas pixels to cover the physical pixels.  That ensures the canvas is completely filled, with potentially a fractional pixel not displayed, but all physical pixels covered.